src/python/WMCore/MicroService/DataStructs/MSOutputTemplate.py

```diff
@@ -240,11 +243,31 @@ def updateDoc(self, myDoc):
             return True
         return False
 
-    def setCampMap(self):
+    def setCampMap(self, doc):
         """
         __setCampMap__
+        Provided the request content retrieved from ReqMgr2, build a map
+        of output dataset per campaign.
+        :param doc: doc is meant to be the request dictionary retrieved from ReqMgr2
         """
-        pass
+        campOutputMap = {}
+        finalMap = []
+        if doc["RequestType"] in ["StepChain", "TaskChain"]:
+            for key, data in doc["ChainParentageMap"]:
+                # key is Step1, Step2 or Task1, Task2, etc
+                # use the Task/Step level campaign, fallback to the top level one
+                campName = doc[key].get("Campaign", doc.get("Campaign"))
+                campOutputMap.setdefault(campName, [])
+                campOutputMap[campName].extend(doc["ChildDsets"])
+            # now just write the final data structure out
+            for campName, dsetList in campOutputMap.items():
+                finalMap.append(dict(campaignName=campName, datasets=dsetList))
+        else:
+            # ReReco and StoreResults
+            finalMap.append(dict(campaignName=doc["Campaign"], datasets=doc["OutputDatasets"]))
+
+        # finally, update this object
+        self.update({"campaignOutputMap": finalMap})
```

Comment:
Please do not go for direct update of the object without passing the regular check of self consistency. Please use the `_checkAttr()`  internal method in a way similar to [1] if you plan to have this as a 'public' method exposed to the outer world. And always return `True` or `False` as an outcome of the function. The Alternative is if you plan this one to be used only as an internal method and rely on the following similar check that is happening in the constructor later [2], then please rename it to `_setCampMap()`. But still return a value from the function. We may also consider renaming the parameter to `myDoc` so to be more clear that while in the function we are referring to the document passed as an argument to the function but not to any other (just for better readability).



[1]
https://github.com/dmwm/WMCore/pull/9719/commits/fca0a0893a7d7ac719a8e8c9d9a53ddd6ecf118a#diff-40f3495cc16a13ae8fd7449995945ce1R222

[2]
https://github.com/dmwm/WMCore/pull/9719/commits/fca0a0893a7d7ac719a8e8c9d9a53ddd6ecf118a#diff-40f3495cc16a13ae8fd7449995945ce1R121

Reply:
I'm going to use the "private" method name then. About the boolean outcome, I think it makes the code less readable and unnecessarily more complex. I'm in favor of simply enforcing a schema here (or raise exceptions upstream). Anyways, this can be followed up in the coming days/weeks.

Reply:
This seems reasonable to me too.